designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,209 @@
+
+For instance: work queue, reconciliation, performance, webhook.
+
+As a result, the layout given by the plugin can be a folder of `config/grafana/`:
+
+```shell
+config/grafana
+├── controller-dashboard.json # The main dashboard raw json file for the simplest usage(copy/paste)
+├── options
+│  ├── performance.json       # CPU/Memory usaage, restful api performance
+│  ├── reconcilier.json       # Reconciliation instruments
+│  ├── webhook.jon            # Webhook Ops
+│  └── workqueue.json         # Work queue status
+└── README.md                 # Docs for the usage
+```
+
+The plugin is triggered to do the scaffolding work for the layout above when the following commands are executed:
+
+- `kubebuilder init -–plugins=grafana/alpha-v1`: when initializing a new project
+- `kubebuilder edit -–plugins=grafana/alpha-v1`: when adding features to an existing project
+
+#### Dashboard Panels
+
+The creation of the panels inside the dashboard can be referred to default [controller-runtime metrics](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/reference/metrics-reference.md).
+
+The overall dashboard should reflect:
+
+1. Latency: The duration of a request for certain controllers/operators, which includes both successful ops and errors ops.
+2. Traffic: The measure of how busy the controllers/operators can be.
+3. Errors: Error rates, or the operations/requests that takes unexpectedly long time.
+4. Saturation: The measure of the possibility of overflows such as deep work queue depth, concurrencies of reconciliations, large cpu/memory usage.
+
+Also, the dashboard should provide filters and selectors to focus on certain objects. This can be implemented by utilizing [Grafana Query Variables](https://grafana.com/docs/grafana/latest/datasources/prometheus/#query-variable).
+
+An example can be:
+![sample-dash](https://doc-0c-bg-docs.googleusercontent.com/docs/securesc/op3r6mi7tplp3i74fcm5ur03mtb42v46/84bkk8a1kontsl4d7ehvrkgs1920s6j5/1654557000000/13713818693191747359/13713818693191747359/1F0eM6qYGCiZZ3dqYOKUG11qxqFSDg9kN?e=view&ax=ACxEAsbPfXom32Dkn-LdACjf86ktIPBQRx3OYuF5l9aSUjEYKcraRYHab1nPuy1VXgDu070T-Qe1VcWWa8ItghTzaP8m39BQJ1ATqVgbdPRlLGcjA8vnjTt_OiRekntAPtedbfLEmCNTSCEDuWK0UBGnln4WbKxiAf7WyOUCRqiy47IWRpUwYXgFIfz4kH6vDoVKaWj8CkFFAKWxnbUz22uQayoK2jwgnFDcB_2EnYiXn8c85tMi9_X9goNyQW64Zi1phrlFvnLlJq6x_bHOdtWZ7mirnoRbu3vmgE0zQezA8rCV1wzTTsfE-161UAdDAawjwaKCFDlyibS-4FTHT0GzZIXu-3ehHx1Txc5YyHpUOnadeSIy6G1qGGY5lzDOk7eGJcInn5vSz0c35Go-PBCI5VoyhljtFh9Irr-VhlIYqwsLcbJWPD9o-Zfmq4-nnPNKb-8_zInFsLjAs4Pj2cLMVDBNeGW7CpOyh4oSrtQSMatyTzdYi4ykAG6aW7KHGoUGpsouhwrQ_WaQbModMmm9742VpyVek-4eB-a4-dSlDXqgUYUHAE4jgvHT5ryeCq7qiLnlNjThafot1o0KaxrBJD4DP1S_BOgN5hXQGKcHiSdi0C_90cIFy9JPVRqlx7zEI4FJDpVBRCs6mECKPt2zR_9veb3wUPTbhkO7g7S81YkcfWkM2gzQDXh6E9ybAqxkFcl_Rk_IvbJ8fvrPUUqUuBuwut7EvQAqlPLqKFoLw39ftcBl2BI&authuser=0&nonce=sk8c49m9ttbto&user=13713818693191747359&hash=6ssipjldoqp2pf2qftgot8fec96m9jbu)
```

Comment:
![image](https://user-images.githubusercontent.com/18136486/172271785-d3fa5a59-83d9-4088-8947-de0274f978a1.png)

Reply:
I loved the dashboard, my question here is do we need to build them manually by importing the json content OR have we a way to generate the whole content and just apply that to check its result?

cc/ @Kavinjsir @varshaprasad96

Reply:
Regardless of how it provides the way for user to 'install' dashboards (or auto-installation without manual process), my origin idea is:

Basically, the plugin will scaffold out such json file from static code, which is similar as how `service-monitor` manifest gets generated by [kustomize](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.4.1/pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/monitor.go#L43).

Which means:
1. We will have a stable content of the json file which may get stored by a constant string.
2. The plugin acts similarly as the `kustomize` plugin, which will generate these json file (considered as grafana manifests).
3. To visualize all default metrics, that may lead to multiple constant strings to store these json content.

Reply:
I am not sure if there are some approaches to generate the manifests pragmatically. (Compared with the above static way)

And it seems like drawing panels in the Web UI remains the main methods to build dashboards.

(Or if there are better ways to deal with this, I'd be glad to apply.)

Reply:
HI @Kavinjsir and @varshaprasad96, 

I could check it out. See that with Grafana we can create a config and load the JSON file on it so that we would be able to make a target which when the user executes it the dashboard would be created. 

See the configmap: https://rancher.com/docs/rancher/v2.5/en/monitoring-alerting/guides/persist-grafana/#2-create-a-configmap-using-the-grafana-json-model

That would mean the target would be like: `kubectlc create configmap <dashboard-name> --from-file=/path/to/json/file -n <operator-ns>` OR better yet we scaffold the configMap and we just apply it on the same ns of the Operator. 

So that means:
- The grafana plugin would scaffold the JSON as described here
- Also, it would add a new target that allows the users just run that to create the Dashboard
-The first init PR to add plugin we can only create the JSON file and we can add the target and improve it incrementally in the follow-ups

ASAP I will try to provide further info and suggestions for this EP. 
But you also can see the : https://www.redhat.com/fr/blog/custom-grafana-dashboards-red-hat-openshift-container-platform-4

Reply:
See here : https://www.lisenet.com/2021/install-and-configure-grafana-on-kubernetes/

@Kavinjsir sharing for you to be able to do manual tests as well and see how we need to config.
I need a little time to be able to finish the suggestions on this doc but you can also apply what was recommended already so that we can close a few comments